In [6]:
import pandas as pd
import json

pd.set_option("display.max_columns", 100)
pd.set_option("display.max_rows", 25)

In [7]:
! source ../conf/local.env && curl --request GET \
  --url "https://api.openalex.org/sources?per_page=25&api_key=${NACSOS_OPENALEX__API_KEY}"  > sample_25.json

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1033k  100 1033k    0     0   526k      0  0:00:01  0:00:01 --:--:--  525k


In [8]:
with open("sample_25.json", "r") as file:
    txt = json.load(file)
txt.keys()

dict_keys(['meta', 'results', 'group_by'])

In [9]:
df = pd.DataFrame(txt["results"])
df

,id,issn_l,issn,display_name,host_organization,host_organization_name,host_organization_lineage,works_count,oa_works_count,cited_by_count,summary_stats,is_oa,is_in_doaj,is_in_doaj_since_year,is_high_oa_rate,is_high_oa_rate_since_year,is_in_scielo,is_ojs,is_core,oa_flip_year,first_publication_year,last_publication_year,ids,homepage_url,apc_prices,apc_usd,country_code,societies,alternate_titles,type,topics,topic_share,counts_by_year,works_api_url,updated_date,created_date
0,https://openalex.org/S7407052672,None,[],NIFS,None,None,[],38965892,38965892,0,"{'2yr_mean_citedness': 0.0, 'h_index': 0, 'i10...",False,False,None,False,NaN,False,False,False,NaN,2024.0,2026.0,{'openalex': 'https://openalex.org/S7407052672...,None,None,None,None,[],[],repository,"[{'id': 'https://openalex.org/T12157', 'displa...","[{'id': 'https://openalex.org/T11920', 'displa...","[{'year': 2026, 'works_count': 10025756, 'oa_w...",https://api.openalex.org/works?filter=primary_...,2026-05-02T10:02:47,2025-10-10T00:00:00
1,https://openalex.org/S7407053449,None,[],DiSSCo,None,None,[],12100903,12100903,2,"{'2yr_mean_citedness': 1.6527692189582878e-07,...",False,False,None,False,NaN,False,False,False,NaN,2024.0,2026.0,{'openalex': 'https://openalex.org/S7407053449...,None,None,None,None,[],[],repository,"[{'id': 'https://openalex.org/T12157', 'displa...","[{'id': 'https://openalex.org/T14254', 'displa...","[{'year': 2026, 'works_count': 4743546, 'oa_wo...",https://api.openalex.org/works?filter=primary_...,2026-06-15T10:02:11,2025-10-10T00:00:00
2,https://openalex.org/S4306400562,None,None,Zenodo (CERN European Organization for Nuclear...,https://openalex.org/I67311998,European Organization for Nuclear Research,[https://openalex.org/I67311998],12031474,12031474,944741,"{'2yr_mean_citedness': 0.02264130986997359, 'h...",True,False,None,True,NaN,False,False,False,NaN,1543.0,2026.0,{'openalex': 'https://openalex.org/S4306400562...,https://zenodo.org,[],None,None,[],[],repository,"[{'id': 'https://openalex.org/T13958', 'displa...","[{'id': 'https://openalex.org/T14200', 'displa...","[{'year': 9999, 'works_count': 77, 'oa_works_c...",https://api.openalex.org/works?filter=primary_...,2026-06-15T10:02:11,2016-06-24T00:00:00
3,https://openalex.org/S4377196282,None,None,Figshare,https://openalex.org/I4210132348,Figshare (United Kingdom),[https://openalex.org/I4210132348],10045527,10045526,508662,"{'2yr_mean_citedness': 0.012479478643640862, '...",False,False,None,False,NaN,False,False,False,NaN,2006.0,2021.0,{'openalex': 'https://openalex.org/S4377196282...,None,[],None,None,[],[],repository,"[{'id': 'https://openalex.org/T10015', 'displa...","[{'id': 'https://openalex.org/T10451', 'displa...","[{'year': 2922, 'works_count': 1, 'oa_works_co...",https://api.openalex.org/works?filter=primary_...,2026-06-15T10:02:11,2016-06-24T00:00:00
4,https://openalex.org/S4306525036,None,None,PubMed,https://openalex.org/I1299303238,National Institutes of Health,[https://openalex.org/I1299303238],7891226,675139,46057749,"{'2yr_mean_citedness': 0.5064651195798737, 'h_...",False,False,None,False,NaN,False,False,False,NaN,1400.0,2026.0,{'openalex': 'https://openalex.org/S4306525036...,https://pubmed.ncbi.nlm.nih.gov,[],None,US,[],[],repository,"[{'id': 'https://openalex.org/T14134', 'displa...","[{'id': 'https://openalex.org/T13365', 'displa...","[{'year': 2026, 'works_count': 11840, 'oa_work...",https://api.openalex.org/works?filter=primary_...,2026-06-15T10:02:11,2016-06-24T00:00:00
5,https://openalex.org/S4210203682,0366-4457,[0366-4457],Bulletin of Miscellaneous Information (Royal G...,None,None,[],7181827,97323,550068,"{'2yr_mean_citedness': 0.029509924468645705, '...",False,False,None,False,NaN,False,False,False,NaN,0.0,2026.0,{'openalex': 'https://openalex.org/S4210203682...,http://www.archive.org/details/mobot3175300225...,[],None,GB,[],"[Bulletin of miscellaneous Information, Kew]",journal,"[{'id': 'https://openalex.org/T10346', 'displa...","[{'id': 'https://openalex.org/T13436', 'd

In [10]:
ids = pd.DataFrame.from_dict(df.ids.to_dict(), orient="index")
ids

,openalex,issn_l,issn,mag,wikidata
0,https://openalex.org/S7407052672,None,[],7407052672,None
1,https://openalex.org/S7407053449,None,[],7407053449,None
2,https://openalex.org/S4306400562,None,None,4306400562,https://www.wikidata.org/wiki/Q22661177
3,https://openalex.org/S4377196282,None,None,4377196282,None
4,https://openalex.org/S4306525036,None,None,4306525036,https://www.wikidata.org/entity/Q180686
5,https://openalex.org/S4210203682,0366-4457,[0366-4457],4210203682,https://www.wikidata.org/entity/Q5735532
6,https://openalex.org/S4398183272,None,None,4398183272,None
7,https://openalex.org/S7407053304,None,[],7407053304,None
8,https://openalex.org/S7407053410,None,[],7407053410,None
9,https://openalex.org/S7407056385,None,None,7407056385,None


In [11]:
df.merge(ids, left_index=True, right_index=True)[["id", "openalex", "issn_x", "issn_y", "issn_l_x", "issn_l_y", "mag"]]

,id,openalex,issn_x,issn_y,issn_l_x,issn_l_y,mag
0,https://openalex.org/S7407052672,https://openalex.org/S7407052672,[],[],None,None,7407052672
1,https://openalex.org/S7407053449,https://openalex.org/S7407053449,[],[],None,None,7407053449
2,https://openalex.org/S4306400562,https://openalex.org/S4306400562,None,None,None,None,4306400562
3,https://openalex.org/S4377196282,https://openalex.org/S4377196282,None,None,None,None,4377196282
4,https://openalex.org/S4306525036,https://openalex.org/S4306525036,None,None,None,None,4306525036
5,https://openalex.org/S4210203682,https://openalex.org/S4210203682,[0366-4457],[0366-4457],0366-4457,0366-4457,4210203682
6,https://openalex.org/S4398183272,https://openalex.org/S4398183272,None,None,None,None,4398183272
7,https://openalex.org/S7407053304,https://openalex.org/S7407053304,[],[],None,None,7407053304
8,https://openalex.org/S7407053410,https://openalex.org/S7407053410,[],[],None,None,7407053410
9,https://openalex.org/S7407056385,https://openalex.org/S7407056385,None,None,None,None,7407056385


In [12]:
! source ../conf/local.env && curl --request GET \
  --url "https://api.openalex.org/sources?sample=25&select=ids%2Cdisplay_name%2Chost_organization%2Chost_organization_name%2Chost_organization_lineage&api_key=${NACSOS_OPENALEX__API_KEY}" 

{"meta": {"count": 25, "db_response_time_ms": 16, "page": 1, "per_page": 25, "groups_count": null, "x_query": {"oql": "sources\nsample 25\nreturn ids, name, publisher, host organization name, host organization lineage", "oqo": {"get_rows": "sources", "sample": 25, "select": ["ids", "display_name", "host_organization", "host_organization_name", "host_organization_lineage"]}, "url": "/sources?select=ids,display_name,host_organization,host_organization_name,host_organization_lineage&sample=25"}, "cost_usd": 0.0001}, "results": [{"ids": {"openalex": "https://openalex.org/S4210210345", "issn_l": "2570-3390", "issn": ["2570-3390", "2570-3404"], "mag": "4210210345", "wikidata": null}, "display_name": "Advances in mental health and addiction", "host_organization": "https://openalex.org/P4310319972", "host_organization_name": "Springer International Publishing", "host_organization_lineage": ["https://openalex.org/P4310319972", "https://openalex.org/P4310319965"]}, {"ids": {"openalex": "https://

## Work with full list of sources

In [13]:
oa = pd.read_json("openalex_sources.json")
print(oa.shape)
oa.head()

(283281, 5)


,ids,display_name,host_organization,host_organization_name,host_organization_lineage
0,{'openalex': 'https://openalex.org/S7407052672...,NIFS,None,None,[]
1,{'openalex': 'https://openalex.org/S7407053449...,DiSSCo,None,None,[]
2,{'openalex': 'https://openalex.org/S4306400562...,Zenodo (CERN European Organization for Nuclear...,https://openalex.org/I67311998,European Organization for Nuclear Research,[https://openalex.org/I67311998]
3,{'openalex': 'https://openalex.org/S4377196282...,Figshare,https://openalex.org/I4210132348,Figshare (United Kingdom),[https://openalex.org/I4210132348]
4,{'openalex': 'https://openalex.org/S4306525036...,PubMed,https://openalex.org/I1299303238,National Institutes of Health,[https://openalex.org/I1299303238]


In [14]:
ids = pd.json_normalize(oa["ids"]).add_prefix("id_")
oa = pd.concat([oa.drop(columns=["ids"]), ids], axis=1)
oa.head()

,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn_l,id_issn,id_mag,id_wikidata
0,NIFS,None,None,[],https://openalex.org/S7407052672,None,[],7407052672,None
1,DiSSCo,None,None,[],https://openalex.org/S7407053449,None,[],7407053449,None
2,Zenodo (CERN European Organization for Nuclear...,https://openalex.org/I67311998,European Organization for Nuclear Research,[https://openalex.org/I67311998],https://openalex.org/S4306400562,None,None,4306400562,https://www.wikidata.org/wiki/Q22661177
3,Figshare,https://openalex.org/I4210132348,Figshare (United Kingdom),[https://openalex.org/I4210132348],https://openalex.org/S4377196282,None,None,4377196282,None
4,PubMed,https://openalex.org/I1299303238,National Institutes of Health,[https://openalex.org/I1299303238],https://openalex.org/S4306525036,None,None,4306525036,https://www.wikidata.org/entity/Q180686


In [15]:
oa.head().isna().sum()

display_name                 0
host_organization            2
host_organization_name       2
host_organization_lineage    0
id_openalex                  0
id_issn_l                    5
id_issn                      3
id_mag                       0
id_wikidata                  3
dtype: int64

--> replace empty lists with None

In [16]:
oa.id_issn = oa.id_issn.apply(lambda x: None if (x is None or len(x) == 0) else x)
oa.host_organization_lineage = oa.host_organization_lineage.apply(lambda x: None if (x is None or len(x) == 0) else x)

In [17]:
oa.head()

,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn_l,id_issn,id_mag,id_wikidata
0,NIFS,None,None,None,https://openalex.org/S7407052672,None,None,7407052672,None
1,DiSSCo,None,None,None,https://openalex.org/S7407053449,None,None,7407053449,None
2,Zenodo (CERN European Organization for Nuclear...,https://openalex.org/I67311998,European Organization for Nuclear Research,[https://openalex.org/I67311998],https://openalex.org/S4306400562,None,None,4306400562,https://www.wikidata.org/wiki/Q22661177
3,Figshare,https://openalex.org/I4210132348,Figshare (United Kingdom),[https://openalex.org/I4210132348],https://openalex.org/S4377196282,None,None,4377196282,None
4,PubMed,https://openalex.org/I1299303238,National Institutes of Health,[https://openalex.org/I1299303238],https://openalex.org/S4306525036,None,None,4306525036,https://www.wikidata.org/entity/Q180686


Is `id_mag` always same as `id_openalex`?

In [18]:
len("https://openalex.org/S")

22

In [19]:
(oa.id_openalex.str[22:] == oa.id_mag).all()

True

Yes, it both columns refer to the unique ID of the OpenAlex. However, in **Solr** `source_id`s `id_mag` need an "s" in the beggining of the id. 

### Identify ID columns for Mapping

Let's check some descriptive stats and identify columns that we can use to map with Scopus / Web of Science indeces.

In [20]:
oa.count()

display_name                 283280
host_organization             73287
host_organization_name        73510
host_organization_lineage     73743
id_openalex                  283281
id_issn_l                    166455
id_issn                      166455
id_mag                       283281
id_wikidata                   82149
dtype: int64

In [21]:
oa[["display_name", "id_mag", "id_issn_l", "host_organization"]].nunique()

display_name         272268
id_mag               283281
id_issn_l            166270
host_organization     12172
dtype: int64

Can't use `id_issn_l` as an id because it's not unique.. let's see whether the concatenated `id_issn` is unique.

In [22]:
oa.id_issn.str.join(",").sample(10)

166351              2720-9776
66192                    None
111898    2502-6038,2684-9585
245641                   None
106557                   None
266932                   None
88991     2636-3607,2636-3615
168051    2959-2100,2959-2119
268786                   None
269257                   None
Name: id_issn, dtype: object

In [23]:
oa["id_issn_concat"] = oa.id_issn.str.join(",")
oa.id_issn_concat.nunique() # compared to count 166455

166451

In [24]:
print(f"id_issn_l = count: {oa.id_issn_l.count()}, nunique: {oa.id_issn_l.nunique()}, num duplicates: {oa.id_issn_l.count() - oa.id_issn_l.nunique()}, nulls: {oa.id_issn_l.isna().sum()}")
print(f"id_issn   = count: {oa.id_issn.count()}, nunique: {oa.id_issn_concat.nunique()}, num duplicates:   {oa.id_issn.count() - oa.id_issn_concat.nunique()}, nulls: {oa.id_issn.isna().sum()}")

id_issn_l = count: 166455, nunique: 166270, num duplicates: 185, nulls: 116826
id_issn   = count: 166455, nunique: 166451, num duplicates:   4, nulls: 116826


In [25]:
# percentage missing data
oa.isna().sum() / oa.shape[0]

display_name                 0.000004
host_organization            0.741292
host_organization_name       0.740505
host_organization_lineage    0.739683
id_openalex                  0.000000
id_issn_l                    0.412403
id_issn                      0.412403
id_mag                       0.000000
id_wikidata                  0.710009
id_issn_concat               0.412403
dtype: float64

In [26]:
mask = oa.id_issn_concat.value_counts()[oa.id_issn_concat.value_counts() > 1]
oa.loc[(oa.id_issn_concat.isin(mask.index))].sort_values("id_issn_concat")

,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn_l,id_issn,id_mag,id_wikidata,id_issn_concat
39428,Jahrbuch der Österreichischen Byzantinistik,None,None,None,https://openalex.org/S5406935911,0378-8660,"[0378-8660, 1810-536X]",5406935911,None,"0378-8660,1810-536X"
39685,Jahrbuch der Österreichischen Byzantinistik,None,None,[None],https://openalex.org/S4306514788,0378-8660,"[0378-8660, 1810-536X]",4306514788,http://www.wikidata.org/entity/Q1453509,"0378-8660,1810-536X"
214109,Revista Andina de Estudios Políticos,None,None,None,https://openalex.org/S5406983929,2221-4135,[2221-4135],5406983929,None,2221-4135
238697,Revista andina de estudios políticos.,None,Observatorio de la Región Andina,[None],https://openalex.org/S4404663026,2221-4135,[2221-4135],4404663026,None,2221-4135
40587,Akofena,None,None,None,https://openalex.org/S7407051855,2706-6312,"[2706-6312, 2708-0633]",7407051855,None,"2706-6312,2708-0633"
165231,Akofena revue scientifique des Sciences du Lan...,None,None,None,https://openalex.org/S7407056122,2706-6312,"[2706-6312, 2708-0633]",7407056122,None,"2706-6312,2708-0633"
223013,JSSAL,None,Chemnitz University of Technology,[None],https://openalex.org/S7407051706,2747-5646,[2747-5646],7407051706,None,2747-5646
255231,Journal of Research in Social Sciences and Lan...,None,None,None,https://openalex.org/S5407053510,2747-5646,[2747-5646],5407053510,None,2747-5646


Based on above, it looks like id_issn_l might be the latest ISSN. And the docs say it is "The ISSN-L (linking ISSN) for this source. This is the Canonical External ID for sources." 

So maybe we can use this column to map indeed. However first we need to understand why it has so many more duplicates than the list of all ISSNs used by the source.

In [27]:
mask = oa.id_issn_l.value_counts()[oa.id_issn_l.value_counts() > 1]
mask

id_issn_l
3108-3943    2
3090-5656    2
2714-2612    2
1846-1867    2
3110-0155    2
            ..
3117-3926    2
2523-9732    2
2833-1826    2
3079-9228    2
2981-8230    2
Name: count, Length: 185, dtype: int64

In [28]:
oa.loc[(oa.id_issn_l.isin(mask.index))].sort_values("id_issn_l").head(24)

,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn_l,id_issn,id_mag,id_wikidata,id_issn_concat
232120,Listy filologické = Folia Philologica,None,None,None,https://openalex.org/S5407053357,0024-4457,[0024-4457],5407053357,None,0024-4457
232286,Listy filologické = Folia Philologica,None,None,None,https://openalex.org/S7407056394,0024-4457,"[0024-4457, 2570-9410]",7407056394,None,"0024-4457,2570-9410"
261698,Offa,None,Nomos Verlag,[None],https://openalex.org/S5406927939,0078-3714,[0078-3714],5406927939,None,0078-3714
189913,Offa,None,None,None,https://openalex.org/S5407054541,0078-3714,"[0078-3714, 2510-7933]",5407054541,None,"0078-3714,2510-7933"
162950,Šport revija za teoretična in praktična vpraša...,None,None,None,https://openalex.org/S7407058129,0353-7455,"[0353-7455, 3024-0603]",7407058129,None,"0353-7455,3024-0603"
255017,Šport revija za teoretična in praktična vpraša...,None,None,None,https://openalex.org/S5406934987,0353-7455,[0353-7455],5406934987,None,0353-7455
39428,Jahrbuch der Österreichischen Byzantinistik,None,None,None,https://openalex.org/S5406935911,0378-8660,"[0378-8660, 1810-536X]",5406935911,None,"0378-8660,1810-536X"
39685,Jahrbuch der Österreichischen Byzantinistik,None,None,[None],https://openalex.org/S4306514788,0378-8660,"[0378-8660, 1810-536X]",4306514788,http://www.wikidata.org/entity/Q1453509,"0378-8660,1810-536X"
116883,Prometheus Rivista di studi classici,None,None,None,https://openalex.org/S5407052210,0391-2698,[0391-2698],5407052210,None,0391-2698
116891,Prometheus Rivista di studi classici,None,None,None,https://openalex.org/S7407057212,0391-2698,"[0391-2698, 2281-1044]",7407057212,None,"0391-2698,2281-1044"


Looks like all duplications result from an update, and we can deduplicate by keeping the most recent information:
- if id_wikidata is not empty, choose the value from non-empty version
- if id_issn is longer, keep the longer version
- if host organisation name is not None, keep that one
- if host_organization_lineage is [None], this seems like a more recent way of updating data

From what I can think, the only reason why we would not want to keep the latest row would be needing the additional information to fall back to when we cannot match based on id_issn_l. Therefore before writing a special deduplication that always keeps the more information, rather than what OpenAlex believes to be currently correct - latest -, let's see how much we can match with just the ID.

### Deduplicate Open Alex Data

#### with last row

In [29]:
print(oa.shape)
oa = oa.groupby("id_issn_l").last().reset_index()
oa.shape

(283281, 10)


(166270, 10)

In [30]:
scie = pd.read_csv("../data/citation_indexes/webofsci/Science Citation Index Expanded (SCIE).csv")
print(scie.shape)
scie.head()

(9430, 7)


,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories
0,2D MATERIALS,2053-1583,2053-1583,IOP PUBLISHING LTD,"No.2 The Distillery, Glassfields, Avon Street,...",English,Nanoscience & Nanotechnology | Materials Scien...
1,3 BIOTECH,2190-572X,2190-5738,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Biotechnology & Applied Microbiology
2,3D PRINTING AND ADDITIVE MANUFACTURING,2329-7662,2329-7670,SAGE PUBLICATIONS INC,"2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320",English,"Engineering, Manufacturing | Materials Science..."
3,4OR-A QUARTERLY JOURNAL OF OPERATIONS RESEARCH,1619-4500,1614-2411,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Operations Research & Management Science
4,AAPG BULLETIN,0149-1423,1558-9153,AMER ASSOC PETROLEUM GEOLOGIST,"1444 S BOULDER AVE, PO BOX 979, TULSA, USA, OK...",English,"Geosciences, Multidisciplinary"


### Match coverage

In [31]:
scie2 = scie.merge(
    oa[["id_issn_l", "display_name", "host_organization", "host_organization_name", "id_mag"]], 
    how="left", left_on="ISSN", right_on="id_issn_l")
print(scie2.shape)
scie2.head()

(9430, 12)


,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,id_issn_l,display_name,host_organization,host_organization_name,id_mag
0,2D MATERIALS,2053-1583,2053-1583,IOP PUBLISHING LTD,"No.2 The Distillery, Glassfields, Avon Street,...",English,Nanoscience & Nanotechnology | Materials Scien...,2053-1583,2D Materials,https://openalex.org/P4310320083,IOP Publishing,4210229423
1,3 BIOTECH,2190-572X,2190-5738,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Biotechnology & Applied Microbiology,2190-572X,3 Biotech,https://openalex.org/P4310319900,Springer Science+Business Media,4210190239
2,3D PRINTING AND ADDITIVE MANUFACTURING,2329-7662,2329-7670,SAGE PUBLICATIONS INC,"2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320",English,"Engineering, Manufacturing | Materials Science...",2329-7662,3D Printing and Additive Manufacturing,https://openalex.org/P4310320443,"Mary Ann Liebert, Inc.",4210174904
3,4OR-A QUARTERLY JOURNAL OF OPERATIONS RESEARCH,1619-4500,1614-2411,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Operations Research & Management Science,NaN,NaN,NaN,NaN,NaN
4,AAPG BULLETIN,0149-1423,1558-9153,AMER ASSOC PETROLEUM GEOLOGIST,"1444 S BOULDER AVE, PO BOX 979, TULSA, USA, OK...",English,"Geosciences, Multidisciplinary",0149-1423,AAPG Bulletin,https://openalex.org/P4310316082,American Association of Petroleum Geologists,144020083


In [32]:
scie2.isna().sum() / scie2.shape[0]

Journal title                0.000000
ISSN                         0.060976
eISSN                        0.011877
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.006045
Web of Science Categories    0.000212
id_issn_l                    0.108165
display_name                 0.108165
host_organization            0.151432
host_organization_name       0.151432
id_mag                       0.108165
dtype: float64

In [33]:
scie2[scie2.ISSN.isna()].sample(5)

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,id_issn_l,display_name,host_organization,host_organization_name,id_mag
2237,CONSERVATION SCIENCE AND PRACTICE,NaN,2578-4854,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,Biodiversity Conservation,NaN,NaN,NaN,NaN,NaN
7742,PHYSICAL REVIEW ACCELERATORS AND BEAMS,NaN,2469-9888,AMER PHYSICAL SOC,"ONE PHYSICS ELLIPSE, COLLEGE PK, USA, MD, 2074...",English,"Physics, Nuclear | Physics, Particles & Fields",NaN,NaN,NaN,NaN,NaN
2867,ENVIRONMENTAL HEALTH,NaN,1476-069X,BMC,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",English,"Environmental Sciences | Public, Environmental...",NaN,NaN,NaN,NaN,NaN
2787,ENERGIES,NaN,1996-1073,MDPI,"MDPI AG, Grosspeteranlage 5, BASEL, Switzerlan...",English,Energy & Fuels,NaN,NaN,NaN,NaN,NaN
7865,PLANTS PEOPLE PLANET,NaN,2572-2611,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,Biodiversity Conservation | Plant Sciences | E...,NaN,NaN,NaN,NaN,NaN


In [34]:
scie3 = scie[scie.ISSN.isna()]
print(scie3.shape)
scie3 = scie3.merge(
    oa[["id_issn_l", "display_name", "host_organization", "host_organization_name", "id_mag"]], 
    how="left", left_on="eISSN", right_on="id_issn_l", validate="1:1")
print(scie3.shape)
scie3.isna().sum() / scie3.shape[0]

(575, 7)


(575, 12)


Journal title                0.000000
ISSN                         1.000000
eISSN                        0.000000
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.046957
Web of Science Categories    0.003478
id_issn_l                    0.142609
display_name                 0.142609
host_organization            0.200000
host_organization_name       0.200000
id_mag                       0.142609
dtype: float64

In [35]:
scie3.sample(5)

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,id_issn_l,display_name,host_organization,host_organization_name,id_mag
94,BMC MUSCULOSKELETAL DISORDERS,NaN,1471-2474,BMC,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",English,Orthopedics | Rheumatology,1471-2474,BMC Musculoskeletal Disorders,https://openalex.org/P4310320256,BioMed Central,35439863
86,BMC HEALTH SERVICES RESEARCH,NaN,1472-6963,BMC,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",English,Health Care Sciences & Services,1472-6963,BMC Health Services Research,https://openalex.org/P4310320256,BioMed Central,12898181
275,IEEE TRANSACTIONS ON COMPUTATIONAL BIOLOGY AND...,NaN,2998-4165,IEEE COMPUTER SOC,"10662 LOS VAQUEROS CIRCLE, PO BOX 3014, LOS AL...",NaN,"Computer Science, Interdisciplinary Applicatio...",2998-4165,IEEE Transactions on Computational Biology and...,None,None,5407042751
145,CLINICAL AND TRANSLATIONAL RADIATION ONCOLOGY,NaN,2405-6308,ELSEVIER IRELAND LTD,"ELSEVIER HOUSE, BROOKVALE PLAZA, EAST PARK SHA...",English,"Oncology | Radiology, Nuclear Medicine & Medic...",2405-6308,Clinical and Translational Radiation Oncology,https://openalex.org/P4310320990,Elsevier BV,2898297225
107,BMC ZOOLOGY,NaN,2056-3132,BMC,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",English,Zoology,2056-3132,BMC Zoology,https://openalex.org/P4310320256,BioMed Central,4210239315


Names of Journals seem to match.. Let's fill the rest with string matching

In [36]:
scie4 = scie3[scie3.id_issn_l.isna()][['Journal title', 'ISSN', 'eISSN', 'Publisher name', 'Publisher address',
       'Languages', 'Web of Science Categories']]
scie4["journal_title"] = scie4["Journal title"].str.lower().str.strip()
scie4.head()

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,journal_title
5,ACTA BIOCHIMICA POLONICA,NaN,1734-154X,FRONTIERS MEDIA SA,"AVENUE DU TRIBUNAL FEDERAL 34, LAUSANNE, SWITZ...",English,Biochemistry & Molecular Biology,acta biochimica polonica
6,ACTA BOTANICA MEXICANA,NaN,2448-7589,INST ECOLOGIA AC,"CENTRO REGIONAL DEL BAJIO, APARTADO POSTAL 386...",Spanish,Plant Sciences,acta botanica mexicana
7,ACTA CRYSTALLOGRAPHICA SECTION B-STRUCTURAL SC...,NaN,2052-5206,INT UNION CRYSTALLOGRAPHY,"2 ABBEY SQ, CHESTER, ENGLAND, CH1 2HU",English,"Chemistry, Multidisciplinary | Crystallography",acta crystallographica section b-structural sc...
11,ADDICTION SCIENCE & CLINICAL PRACTICE,NaN,1940-0640,BMC,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",English,Substance Abuse,addiction science & clinical practice
20,AGRARFORSCHUNG SCHWEIZ,NaN,2813-317X,AGRARFORSCHUNG,"AGROSCOPE LIEBEFELD-POSIEUX, POSIEUX, SWITZERL...",Multi-Language,"Agriculture, Multidisciplinary",agrarforschung schweiz


In [37]:
oa["journal_title"] = oa["display_name"].str.lower().str.strip()
unique_title_mask = oa.journal_title.value_counts()[oa.journal_title.value_counts() == 1]
oa2 = oa.loc[(oa.journal_title.isin(unique_title_mask.index))]

print(scie4.shape)
scie4 = scie4.merge(
    oa2[["id_issn_l", "display_name", "host_organization", "host_organization_name", "id_mag", 
        "journal_title"]], 
    how="left", on="journal_title", validate="1:1")
print(scie4.shape)
scie4.isna().sum() / scie4.shape[0]

(82, 8)
(82, 13)


Journal title                0.000000
ISSN                         1.000000
eISSN                        0.000000
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.060976
Web of Science Categories    0.012195
journal_title                0.000000
id_issn_l                    0.317073
display_name                 0.317073
host_organization            0.365854
host_organization_name       0.365854
id_mag                       0.317073
dtype: float64

In [38]:
oa.journal_title.value_counts().head()

journal_title
proceedings                            74
jurnal pengabdian kepada masyarakat    20
jurnal pengabdian masyarakat           15
j-institute                            14
jurnal kesehatan                       13
Name: count, dtype: int64

In [39]:
scie2.isna().sum() / scie2.shape[0]

Journal title                0.000000
ISSN                         0.060976
eISSN                        0.011877
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.006045
Web of Science Categories    0.000212
id_issn_l                    0.108165
display_name                 0.108165
host_organization            0.151432
host_organization_name       0.151432
id_mag                       0.108165
dtype: float64

In [40]:
((scie2.ISSN.notna()) & (scie2.id_mag.isna())).sum()

445

In [41]:
scie2

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,id_issn_l,display_name,host_organization,host_organization_name,id_mag
0,2D MATERIALS,2053-1583,2053-1583,IOP PUBLISHING LTD,"No.2 The Distillery, Glassfields, Avon Street,...",English,Nanoscience & Nanotechnology | Materials Scien...,2053-1583,2D Materials,https://openalex.org/P4310320083,IOP Publishing,4210229423
1,3 BIOTECH,2190-572X,2190-5738,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Biotechnology & Applied Microbiology,2190-572X,3 Biotech,https://openalex.org/P4310319900,Springer Science+Business Media,4210190239
2,3D PRINTING AND ADDITIVE MANUFACTURING,2329-7662,2329-7670,SAGE PUBLICATIONS INC,"2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320",English,"Engineering, Manufacturing | Materials Science...",2329-7662,3D Printing and Additive Manufacturing,https://openalex.org/P4310320443,"Mary Ann Liebert, Inc.",4210174904
3,4OR-A QUARTERLY JOURNAL OF OPERATIONS RESEARCH,1619-4500,1614-2411,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Operations Research & Management Science,NaN,NaN,NaN,NaN,NaN
4,AAPG BULLETIN,0149-1423,1558-9153,AMER ASSOC PETROLEUM GEOLOGIST,"1444 S BOULDER AVE, PO BOX 979, TULSA, USA, OK...",English,"Geosciences, Multidisciplinary",0149-1423,AAPG Bulletin,https://openalex.org/P4310316082,American Association of Petroleum Geologists,144020083
...,...,...,...,...,...,...,...,...,...,...,...,...
9425,ZOOSYSTEMA,1280-9551,1638-9387,"PUBLICATIONS SCIENTIFIQUES DU MUSEUM, PARIS","CP 39-57, RUE CUVIER, PARIS CEDEX 05, FRANCE, ...",English,Zoology,1280-9551,Zoosystema,https://openalex.org/P4310311095,Muséum national d'histoire naturelle,141665667
9426,ZOOSYSTEMATICS AND EVOLUTION,1860-0743,1860-0743,PENSOFT PUBLISHERS,"12 PROF GEORGI ZLATARSKI ST, SOFIA, BULGARIA, ...",Multi-Language,Zoology,NaN,NaN,NaN,NaN,NaN
9427,ZOOTAXA,1175-5326,1175-5334,MAGNOLIA PRESS,"250F Marua Road Mt Wellington, AUCKLAND, New Z...",English,Zoology,1175-5326,Zootaxa,https://openalex.org/P4310321855,Q15088586,171471881
9428,ZUCHTUNGSKUNDE,0044-5401,1867-4518,EUGEN ULMER GMBH CO,"POSTFACH 700561 WOLLGRASWEG 41, STUTTGART, GER...",German,"Agriculture, Dairy & Animal Science",0044-5401,Züchtungskunde,https://openalex.org/P4310319817,Eugen Ulmer Verlag,34199213


In [42]:
scie2[(scie2.ISSN.notna()) & (scie2.id_mag.isna())].count() / scie2.shape[0]

Journal title                0.047190
ISSN                         0.047190
eISSN                        0.046129
Publisher name               0.047190
Publisher address            0.047190
Languages                    0.046023
Web of Science Categories    0.047190
id_issn_l                    0.000000
display_name                 0.000000
host_organization            0.000000
host_organization_name       0.000000
id_mag                       0.000000
dtype: float64

with 2 falllbacks to eISSN and titles, we get only %30 of 82 rows unmatched --> 25 rows out of original 9430 - 445  --> 0.003 missing. Maybe that can be recovered with string beginning-ending matching..

In [43]:
scie4.sample(5)

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,journal_title,id_issn_l,display_name,host_organization,host_organization_name,id_mag
29,ETHICS AND SOCIETY,NaN,3091-3322,SPRINGER,"VAN GODEWIJCKSTRAAT 30, DORDRECHT, NETHERLANDS...",NaN,NaN,ethics and society,3091-3314,Ethics and Society,None,None,5407055850
35,GEOARCHAEOLOGY-AN INTERNATIONAL JOURNAL,NaN,1520-6548,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,"Geosciences, Multidisciplinary",geoarchaeology-an international journal,NaN,NaN,NaN,NaN,NaN
20,CLINICAL INTERVENTIONS IN AGING,NaN,1178-1998,DOVE MEDICAL PRESS LTD,"PO BOX 300-008, ALBANY, NEW ZEALAND, AUCKLAND,...",English,Geriatrics & Gerontology,clinical interventions in aging,1176-9092,Clinical Interventions in Aging,https://openalex.org/P4310320348,Dove Medical Press,68257898
70,PSYCHIATRY INVESTIGATION,NaN,1976-3026,KOREAN NEUROPSYCHIATRIC ASSOC,"RN 522, G-FIVE CENTRAL PLAZA 1685-8 SEOCHO 4-D...",English,Psychiatry,psychiatry investigation,1738-3684,Psychiatry Investigation,https://openalex.org/P4365920587,Korean Neuropsychiatric Association,121704485
2,ACTA CRYSTALLOGRAPHICA SECTION B-STRUCTURAL SC...,NaN,2052-5206,INT UNION CRYSTALLOGRAPHY,"2 ABBEY SQ, CHESTER, ENGLAND, CH1 2HU",English,"Chemistry, Multidisciplinary | Crystallography",acta crystallographica section b-structural sc...,NaN,NaN,NaN,NaN,NaN


In [44]:
row = scie4[scie4.id_mag.isna()].sample()
title = row.journal_title.iloc[0]
print(row)
issn = row.ISSN.iloc[0]
if pd.isna(issn):
    issn = row.eISSN.iloc[0]
print(issn)
if pd.isna(issn):
    result = oa2[(oa2.journal_title.str.contains(title))]
else:
    result = oa2[(oa2.journal_title.str.contains(title)) | (oa2.id_issn_concat.str.contains(issn))]
result

                          Journal title ISSN      eISSN  \
6  ANNALS ACADEMY OF MEDICINE SINGAPORE  NaN  2972-4066   

            Publisher name                                  Publisher address  \
6  ACAD MEDICINE SINGAPORE  81 Kim Keat Road, #12-00 NKF Centre, Singapore...   

  Languages     Web of Science Categories  \
6   English  Medicine, General & Internal   

                          journal_title id_issn_l display_name  \
6  annals academy of medicine singapore       NaN          NaN   

  host_organization host_organization_name id_mag  
6               NaN                    NaN    NaN  
2972-4066


,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat,journal_title


In [45]:
title = "transfusion"
row = scie4[scie4.journal_title == title]
print(row)
issn = row.ISSN.iloc[0]
if pd.isna(issn):
    issn = row.eISSN.iloc[0]
print(issn)
if pd.isna(issn):
    result = oa2[(oa2.journal_title.str.contains(title))]
else:
    result = oa2[(oa2.journal_title.str.contains(title)) | (oa2.id_issn_concat.str.contains(issn))]
result

   Journal title ISSN      eISSN Publisher name  \
78   TRANSFUSION  NaN  1537-2995          WILEY   

                             Publisher address Languages  \
78  111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774   English   

   Web of Science Categories journal_title id_issn_l display_name  \
78                Hematology   transfusion       NaN          NaN   

   host_organization host_organization_name id_mag  
78               NaN                    NaN    NaN  
1537-2995


,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat,journal_title
6578,0035-2977,Revue Française de Transfusion,https://openalex.org/P4310320990,Elsevier BV,[https://openalex.org/P4310320990],https://openalex.org/S4210180146,"[0035-2977, 2213-946X]",4210180146,None,"0035-2977,2213-946X",revue française de transfusion
9357,0074-8528,Proceedings of the International Society of Bl...,None,None,None,https://openalex.org/S4210225101,[0074-8528],4210225101,None,0074-8528,proceedings of the international society of bl...
14255,0171-7111,Haematology and blood transfusion,https://openalex.org/P4310319965,Springer Nature,[https://openalex.org/P4310319965],https://openalex.org/S2755861557,[0171-7111],2755861557,https://www.wikidata.org/entity/Q27719500,0171-7111,haematology and blood transfusion
18432,0278-6222,Plasma Therapy and Transfusion Technology,https://openalex.org/P4310320990,Elsevier BV,[https://openalex.org/P4310320990],https://openalex.org/S183557914,"[0278-6222, 1878-2019]",183557914,http://www.wikidata.org/entity/Q96328899,"0278-6222,1878-2019",plasma therapy and transfusion technology
18993,0301-3553,Ergebnisse der Bluttransfusionsforschung,None,None,None,https://openalex.org/S4210223732,[0301-3553],4210223732,http://www.wikidata.org/entity/Q93827960,0301-3553,ergebnisse der bluttransfusionsforschung
...,...,...,...,...,...,...,...,...,...,...,...
113314,2575-8934,Open Access Blood Research & Transfusion Journal,None,None,None,https://openalex.org/S4210221639,[2575-8934],4210221639,https://www.wikidata.org/entity/Q96721458,2575-8934,open access blood research & transfusion journal
149601,2978-8250,Transplant and Transfusion,None,None,None,https://openalex.org/S7407057746,[2978-8250],7407057746,None,2978-8250,transplant and transfusion
150012,2985-2390,Journal of Hematology and Transfusion Medicine,None,None,None,https://openalex.org/S4404677111,"[2985-2390, 2985-2404]",4404677111,None,"2985-2390,2985-2404",journal of hematology and transfusion medicine
161009,3070-1937,International Journal of Blood Transfusion,None,None,None,https://openalex.org/S5407054750,[3070-1937],5407054750,None,3070-1937,international journal of blood transfusion


!! with sting, we can accidentally match with something wrong.

#### Just with ID matching

So strategy should be
1) if ISSN, match with ISSN
2) if left has ISSN but right does not, match with other entry in ISSN list from right
3) left eISSn -> right issn_l
4) if left eISSN and right null -> check ISSN list from right

In [46]:
oa.shape

(166270, 11)

In [47]:
oa.head()

,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat,journal_title
0,,GEOTROPE,None,None,None,https://openalex.org/S7407051494,"[, 1817-5589, 2960-2084]",7407051494,None,",1817-5589,2960-2084",geotrope
1,0000-0205,Tropical Parasitology,https://openalex.org/P4310320448,Medknow,"[https://openalex.org/P4310320448, https://ope...",https://openalex.org/S4210234080,[0000-0205],4210234080,https://www.wikidata.org/entity/Q96738528,0000-0205,tropical parasitology
2,0000-1112,Journal of Musculoskeletal Diseases in Ayurveda,https://openalex.org/P4310320448,Medknow,"[https://openalex.org/P4310320448, https://ope...",https://openalex.org/S4387277910,[0000-1112],4387277910,None,0000-1112,journal of musculoskeletal diseases in ayurveda
3,0000-992X,Andrew Gilmartin Review,None,None,None,https://openalex.org/S5406922747,[0000-992X],5406922747,None,0000-992X,andrew gilmartin review
4,0001-0197,A A R N newsletter,None,None,None,https://openalex.org/S4306499931,[0001-0197],4306499931,http://www.wikidata.org/entity/Q27708670,0001-0197,a a r n newsletter


In [48]:
oa.id_issn.apply(len).value_counts()

id_issn
1    87643
2    77303
3     1207
4       96
5       17
7        2
6        2
Name: count, dtype: int64

In [49]:
oa.id_issn.head().apply(lambda x: [i for i in x if i]) #.apply(len).value_counts()

0    [1817-5589, 2960-2084]
1               [0000-0205]
2               [0000-1112]
3               [0000-992X]
4               [0001-0197]
Name: id_issn, dtype: object

In [50]:
oa.id_issn.apply(lambda x: [i for i in x if i]).apply(len).value_counts()

id_issn
1    87643
2    77304
3     1206
4       96
5       17
7        2
6        2
Name: count, dtype: int64

In [51]:
oa.sample(5)

,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat,journal_title
10809,0096-946X,American Documentation,https://openalex.org/P4310320595,Wiley,[https://openalex.org/P4310320595],https://openalex.org/S206754237,"[0096-946X, 1936-6108]",206754237,https://www.wikidata.org/entity/Q53952379,"0096-946X,1936-6108",american documentation
99892,2398-0281,Journal of Diabetology and Endocrinology,https://openalex.org/P4310311864,NobleResearch Publishers,[https://openalex.org/P4310311864],https://openalex.org/S4210202477,[2398-0281],4210202477,https://www.wikidata.org/entity/Q96713529,2398-0281,journal of diabetology and endocrinology
109218,2524-6194,Contemporary Chinese Civil and Commercial Law,None,None,None,https://openalex.org/S4393918779,"[2524-6194, 2524-6208]",4393918779,None,"2524-6194,2524-6208",contemporary chinese civil and commercial law
12669,0141-6421,Journal of Petroleum Geology,https://openalex.org/P4310320595,Wiley,[https://openalex.org/P4310320595],https://openalex.org/S198628674,"[0141-6421, 1747-5457]",198628674,None,"0141-6421,1747-5457",journal of petroleum geology
132925,2731-085X,Journal of Rare Diseases,https://openalex.org/P4310319900,Springer Science+Business Media,"[https://openalex.org/P4310319900, https://ope...",https://openalex.org/S4387285336,[2731-085X],4387285336,None,2731-085X,journal of rare diseases


Is `id_issn_l` alwyays the first in `id_issn` list? --> yes.

In [52]:
(oa.id_issn_l == oa.id_issn.apply(lambda x: x[0])).sum() / oa.shape[0]

1.0

In [53]:
scie["ISSN_all"] = scie[["ISSN", "eISSN"]].apply(lambda r: [v for v in r.tolist() if pd.notna(v)], axis=1)
scie["ISSN_filled"] = scie.ISSN
scie.loc[scie.ISSN_filled.isna(), "ISSN_filled"] = scie.loc[scie.ISSN_filled.isna(), "eISSN"]

scie.head()

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_all,ISSN_filled
0,2D MATERIALS,2053-1583,2053-1583,IOP PUBLISHING LTD,"No.2 The Distillery, Glassfields, Avon Street,...",English,Nanoscience & Nanotechnology | Materials Scien...,"[2053-1583, 2053-1583]",2053-1583
1,3 BIOTECH,2190-572X,2190-5738,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Biotechnology & Applied Microbiology,"[2190-572X, 2190-5738]",2190-572X
2,3D PRINTING AND ADDITIVE MANUFACTURING,2329-7662,2329-7670,SAGE PUBLICATIONS INC,"2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320",English,"Engineering, Manufacturing | Materials Science...","[2329-7662, 2329-7670]",2329-7662
3,4OR-A QUARTERLY JOURNAL OF OPERATIONS RESEARCH,1619-4500,1614-2411,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Operations Research & Management Science,"[1619-4500, 1614-2411]",1619-4500
4,AAPG BULLETIN,0149-1423,1558-9153,AMER ASSOC PETROLEUM GEOLOGIST,"1444 S BOULDER AVE, PO BOX 979, TULSA, USA, OK...",English,"Geosciences, Multidisciplinary","[0149-1423, 1558-9153]",0149-1423


In [54]:
def merge_by_bidirectional_preference(dfL, dfR,
                                      left_id_col='left_id', left_list_col='left_keys',
                                      right_id_col='right_id', right_list_col='right_keys',
                                      right_keep_cols=None,
                                      score_weights=(1.0, 1.0),
                                      ):
    """
    Merge dfL and dfR when both sides have ordered lists of alternative keys.
    Returns (result_df, stats) where result_df has one best match per left row (or NaNs if no match),
    and stats is a Series counting matches by chosen left_pref (l_pref) and right_pref (r_pref).

    Parameters:
    - dfL: left DataFrame. Must contain left_id_col and left_list_col (list-like per row).
    - dfR: right DataFrame. Must contain right_id_col and right_list_col (list-like per row).
    - left_id_col/right_id_col: identifier column names for left/right (must be present).
    - left_list_col/right_list_col: names of columns containing lists of alternatives.
    - right_keep_cols: list of columns from dfR to include in result (None => all except right_list_col).
    - score_weights: (wL, wR) weights for scoring: score = wL*l_pref + wR*r_pref.
    """
    tie_breaker=('score','l_pref','r_pref', right_id_col)

    # --- Prepare left exploded ---
    dfL_e = dfL.reset_index(drop=False).rename(columns={'index':'_orig_index'})  # preserve insertion order
    if left_id_col not in dfL_e:
        dfL_e[left_id_col] = dfL_e['_orig_index']
    dfL_e = dfL_e.explode(left_list_col).reset_index(drop=True)
    dfL_e['_lidx'] = dfL_e.groupby('_orig_index').cumcount()  # l_pref
    dfL_e = dfL_e.rename(columns={left_list_col: '_key_value'})
    dfL_e = dfL_e.rename(columns={'_lidx':'l_pref'})

    # --- Prepare right exploded ---
    dfR_e = dfR.reset_index(drop=False).rename(columns={'index':'_r_orig_index'})
    if right_id_col not in dfR_e:
        dfR_e[right_id_col] = dfR_e['_r_orig_index']
    dfR_e = dfR_e.explode(right_list_col).reset_index(drop=True)
    dfR_e['_ridx'] = dfR_e.groupby('_r_orig_index').cumcount()  # r_pref
    dfR_e = dfR_e.rename(columns={right_list_col: '_key_value', '_ridx':'r_pref'})

    # Which right columns to keep
    if right_keep_cols is None:
        right_keep_cols = [c for c in dfR_e.columns if c not in ('_key_value','r_pref','_r_orig_index')]
    right_keep_cols = list(dict.fromkeys(right_keep_cols))  # unique preserve order

    # --- Merge on key value ---
    merged = dfL_e.merge(dfR_e[[ '_key_value','r_pref','_r_orig_index'] + right_keep_cols ],
                         on='_key_value', how='left', suffixes=('_L','_R'))

    # --- Score & tie-break ---
    wL, wR = score_weights
    merged['l_pref'] = merged['l_pref'].astype(float)
    merged['r_pref'] = merged['r_pref'].astype(float)
    merged['score'] = wL * merged['l_pref'] + wR * merged['r_pref']

    # Sort for deterministic choice
    sort_cols = list(tie_breaker)
    # Ensure all tie_breaker cols exist in merged; if not, fallback to score,l_pref,r_pref
    for c in sort_cols:
        if c not in merged.columns:
            sort_cols = ['score','l_pref','r_pref']
            break
    merged_sorted = merged.sort_values(by=sort_cols)

    # Pick best per original left row (use _orig_index to map back)
    best = merged_sorted.drop_duplicates(subset=['_orig_index'], keep='first')

    # Build result: left original columns + chosen right columns + matching metadata
    left_cols = [c for c in dfL.columns if c not in (left_list_col,)]
    out_cols = left_cols + right_keep_cols + ['l_pref','r_pref','score','_key_value']
    result = best.reindex(columns=out_cols).reset_index(drop=True)

    # Stats: counts by l_pref and r_pref where match exists
    stats = {
        'total_left_rows': dfL.shape[0],
        'matched_rows': int(result[right_id_col].notna().sum()),
        'unmatched_rows': int(result[right_id_col].isna().sum())
    }
    # distribution by l_pref and r_pref
    stats['matched_by_l_pref'] = result.loc[result[right_id_col].notna(), 'l_pref'].value_counts().sort_index()
    stats['matched_by_r_pref'] = result.loc[result[right_id_col].notna(), 'r_pref'].value_counts().sort_index()

    return result, stats

In [55]:
print("Left shape:  ", scie.shape)
print("Right shape: ", oa.shape)
m, s = merge_by_bidirectional_preference(scie, oa, left_id_col='ISSN_filled', left_list_col='ISSN_all',
                                      right_id_col='id_issn_l', right_list_col='id_issn')
print("Merged shape:", m.shape)
m.isna().sum() / m.shape[0]

Left shape:   (9430, 9)
Right shape:  (166270, 11)
Merged shape: (9430, 22)


Journal title                0.000000
ISSN                         0.060976
eISSN                        0.011877
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.006045
Web of Science Categories    0.000212
ISSN_filled                  0.000000
id_issn_l                    0.002651
display_name                 0.002651
host_organization            0.054295
host_organization_name       0.054295
host_organization_lineage    0.054189
id_openalex                  0.002651
id_mag                       0.002651
id_wikidata                  0.195758
id_issn_concat               0.002651
journal_title                0.002651
l_pref                       0.000000
r_pref                       0.002651
score                        0.002651
_key_value                   0.000000
dtype: float64

In [56]:
s

{'total_left_rows': 9430,
 'matched_rows': 9405,
 'unmatched_rows': 25,
 'matched_by_l_pref': l_pref
 0.0    9371
 1.0      34
 Name: count, dtype: int64,
 'matched_by_r_pref': r_pref
 0.0    8935
 1.0     447
 2.0      23
 Name: count, dtype: int64}

In [57]:
scie.merge(oa, how='left', left_on='ISSN_filled', right_on='id_issn_l').isna().sum()

Journal title                   0
ISSN                          575
eISSN                         112
Publisher name                  0
Publisher address               0
Languages                      57
Web of Science Categories       2
ISSN_all                        0
ISSN_filled                     0
id_issn_l                     527
display_name                  527
host_organization             968
host_organization_name        968
host_organization_lineage     967
id_openalex                   527
id_issn                       527
id_mag                        527
id_wikidata                  2212
id_issn_concat                527
journal_title                 527
dtype: int64

In [58]:
m.isna().sum()

Journal title                   0
ISSN                          575
eISSN                         112
Publisher name                  0
Publisher address               0
Languages                      57
Web of Science Categories       2
ISSN_filled                     0
id_issn_l                      25
display_name                   25
host_organization             512
host_organization_name        512
host_organization_lineage     511
id_openalex                    25
id_mag                         25
id_wikidata                  1846
id_issn_concat                 25
journal_title                  25
l_pref                          0
r_pref                         25
score                          25
_key_value                      0
dtype: int64

In [59]:
m.head()

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_filled,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_mag,id_wikidata,id_issn_concat,journal_title,l_pref,r_pref,score,_key_value
0,COMMUNICATIONS OF THE ACM,0001-0782,1557-7317,ASSOC COMPUTING MACHINERY,"1601 Broadway, 10th Floor, NEW YORK, Usa, NY, ...",English,"Computer Science, Hardware & Architecture | Co...",0001-0782,0001-0782,Communications of the ACM,https://openalex.org/P4310319798,Association for Computing Machinery,[https://openalex.org/P4310319798],https://openalex.org/S103482838,103482838,https://www.wikidata.org/entity/Q1120519,"0001-0782,1557-7317",communications of the acm,0.0,0.0,0.0,0001-0782
1,AIAA JOURNAL,0001-1452,1533-385X,AMER INST AERONAUTICS ASTRONAUTICS,"1801 ALEXANDER BELL DRIVE, STE 500, RESTON, US...",English,"Engineering, Aerospace",0001-1452,0001-1452,AIAA Journal,https://openalex.org/P4310315709,American Institute of Aeronautics and Astronau...,[https://openalex.org/P4310315709],https://openalex.org/S53336109,53336109,https://www.wikidata.org/entity/Q4651795,"0001-1452,1081-0102,1533-385X",aiaa journal,0.0,0.0,0.0,0001-1452
2,AICHE JOURNAL,0001-1541,1547-5905,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,"Engineering, Chemical",0001-1541,0001-1541,AIChE Journal,https://openalex.org/P4310320595,Wiley,[https://openalex.org/P4310320595],https://openalex.org/S147177414,147177414,https://www.wikidata.org/entity/Q15753322,"0001-1541,1547-5905",aiche journal,0.0,0.0,0.0,0001-1541
3,AORN JOURNAL,0001-2092,1878-0369,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,Nursing,0001-2092,0001-2092,AORN Journal,https://openalex.org/P4310320595,Wiley,[https://openalex.org/P4310320595],https://openalex.org/S5303612,5303612,https://www.wikidata.org/entity/Q4653156,"0001-2092,1878-0369",aorn journal,0.0,0.0,0.0,0001-2092
4,ASHRAE JOURNAL,0001-2491,1943-6637,AMER SOC HEATING REFRIGERATING AIR-CONDITIONIN...,"1791 TULLIE CIRCLE NE, ATLANTA, USA, GA, 30329",English,Thermodynamics | Construction & Building Techn...,0001-2491,0001-2491,ASHRAE journal,https://openalex.org/P4310320990,Elsevier BV,[https://openalex.org/P4310320990],https://openalex.org/S1170834,1170834,https://www.wikidata.org/entity/Q15751177,0001-2491,ashrae journal,0.0,0.0,0.0,0001-2491


In [60]:
# sanity check
m[m.ISSN !=  m.id_issn_l][["ISSN", "eISSN", "_key_value", "id_issn_l", "id_issn_concat", "Journal title", "display_name" ]].dropna()

,ISSN,eISSN,_key_value,id_issn_l,id_issn_concat,Journal title,display_name
8904,0236-5294,1588-2632,0236-5294,0001-5954,"0001-5954,0236-5294,1588-2632",ACTA MATHEMATICA HUNGARICA,Acta Mathematica Academiae Scientiarum Hungaricae
8906,1612-9199,1612-9199,1612-9199,0003-9098,"0003-9098,1612-9199",EUROPEAN POULTRY SCIENCE,Archiv für Geflügelkunde
8907,1464-7931,1469-185X,1464-7931,0006-3231,"0006-3231,1464-7931,1469-185X",BIOLOGICAL REVIEWS,Biological reviews/Biological reviews of the C...
8908,1341-1357,1881-7122,1341-1357,0007-5124,"0007-5124,1341-1357,1881-7122,1884-4162,1884-4170",EXPERIMENTAL ANIMALS,EXPERIMENTAL ANIMALS
8909,2245-7070,2245-7070,2245-7070,0011-6297,"0011-6297,2245-7070",BULLETIN OF THE GEOLOGICAL SOCIETY OF DENMARK,Bulletin of the Geological Society of Denmark
...,...,...,...,...,...,...,...
9400,1735-0328,1726-6890,1735-0328,1726-6882,"1726-6882,1726-6890,1735-0328",IRANIAN JOURNAL OF PHARMACEUTICAL RESEARCH,Iranian journal of pharmaceutical research
9401,1749-4877,1749-4869,1749-4877,1749-4868,"1749-4868,1749-4869,1749-4877",INTEGRATIVE ZOOLOGY,Integrative Zoology
9402,2468-0133,2468-0133,2468-0133,2096-0522,"2096-0522,2096-675X,2468-0133",JOURNAL OF OCEAN ENGINEERING AND SCIENCE,Journal of Ocean Engineering and Science
9403,2225-6253,2411-9717,2411-9717,0038-223X,"0038-223X,2411-9717",JOURNAL OF THE SOUTHERN AFRICAN INSTITUTE OF M...,Journal of the Southern African Institute of M...


In [61]:
m[m.ISSN !=  m.id_issn_l][["ISSN", "eISSN", "_key_value", "id_issn_l", "id_issn_concat", "Journal title", "display_name" ]].dropna().sample(25)

,ISSN,eISSN,_key_value,id_issn_l,id_issn_concat,Journal title,display_name
9321,2560-6921,2560-6778,2560-6921,2560-6778,"2560-6778,2560-6921",JOURNAL OF NONLINEAR AND VARIATIONAL ANALYSIS,Journal of Nonlinear and Variational Analysis
9225,2090-8083,2042-0080,2090-8083,2042-0080,"2042-0080,2090-8083",PARKINSONS DISEASE,Parkinson s Disease
9189,1877-0320,1877-0320,1877-0320,1877-0312,"1877-0312,1877-0320",REVUE FRANCAISE D ALLERGOLOGIE,Revue française d'allergologie
9181,1876-3413,1876-3405,1876-3413,1876-3405,"1876-3405,1876-3413",INTERNATIONAL HEALTH,International Health
9338,2724-6507,2724-6116,2724-6507,2724-6116,"2724-6116,2724-6507",MINERVA ENDOCRINOLOGY,Minerva Endocrinology
9320,2543-1536,2542-6605,2543-1536,2542-6605,"2542-6605,2543-1536",INTERNET OF THINGS,Internet of Things
9050,1479-3261,1476-4431,1479-3261,1476-4431,"1476-4431,1479-3261",JOURNAL OF VETERINARY EMERGENCY AND CRITICAL CARE,Journal of Veterinary Emergency and Critical Care
9240,2090-1232,2090-1224,2090-1232,2090-1224,"2090-1224,2090-1232",JOURNAL OF ADVANCED RESEARCH,Journal of Advanced Research
9352,1002-185X,1875-5372,1875-5372,1875-5372,1875-5372,RARE METAL MATERIALS AND ENGINEERING,Rare Metal Materials and Engineering
9202,1937-5093,1937-5077,1937-5093,1937-5077,"1937-5077,1937-5093",KINETIC AND RELATED MODELS,Kinetic and Related Models


ok, matching from older ISSNs in OpenAlex or switching to eISSN when ISSN is not found in web of science index seems to match reasonably based on the titles of the journals. We have only 0.002651 missing (25 rows). Let's check what's still missing for the last time.

In [62]:
m[m.id_issn_l.isna()]

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_filled,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_mag,id_wikidata,id_issn_concat,journal_title,l_pref,r_pref,score,_key_value
9405,ACTA PHYTOTAXONOMICA ET GEOBOTANICA,1346-7565,2189-7042,"JAPANESE SOC PLANT SYSTEMATICS, UNIV TOKYO, GR...","BOTANICAL GARDENS, 3-7-1 HAKUSAN, TOKYO, JAPAN...",English,Plant Sciences,1346-7565,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,1346-7565
9406,AGRARFORSCHUNG SCHWEIZ,NaN,2813-317X,AGRARFORSCHUNG,"AGROSCOPE LIEBEFELD-POSIEUX, POSIEUX, SWITZERL...",Multi-Language,"Agriculture, Multidisciplinary",2813-317X,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,2813-317X
9407,ANNALI DI BOTANICA,0365-0812,2239-3129,UNIV STUDI ROMA LA SAPIENZA,"PIAZZALE ALDO MORO, 5, ROME, ITALY, I-00185",Multi-Language,Plant Sciences | Environmental Sciences,0365-0812,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,0365-0812
9408,ANNALS ACADEMY OF MEDICINE SINGAPORE,NaN,2972-4066,ACAD MEDICINE SINGAPORE,"81 Kim Keat Road, #12-00 NKF Centre, Singapore...",English,"Medicine, General & Internal",2972-4066,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,2972-4066
9409,ASIAN MYRMECOLOGY,1985-1944,NaN,UNIV MALAYSIA SABAH,"JALAN UMS, 88400, KOTA KINABALU, MALAYSIA, SAB...",English,Entomology,1985-1944,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,1985-1944
9410,ATW-INTERNATIONAL JOURNAL FOR NUCLEAR POWER,1431-5254,NaN,INFORUM VERLAGS-VERWALTUNGSGESELLSCHAFT MBH,"ATW-ADVERTISING-SUBSCRIPTIONS, ROBERT-KOCH-PLA...",Multi-Language,Nuclear Science & Technology,1431-5254,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,1431-5254
9411,AUSTRIAN JOURNAL OF FOREST SCIENCE,0379-5292,0375-524X,OSTERREICHISCHER AGRARVERLAG,"BANKGASSE 13, VIENNA 1, AUSTRIA, 1014",German,Forestry,0379-5292,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,0379-5292
9412,BERICHTE UBER LANDWIRTSCHAFT,2196-5099,NaN,BUNDESMINISTERIUM ERNAHRUNG LANDWIRTSCHAFT,"DEICHMANNS AUE 29, BPMM, GERMANY, 53179",Multi-Language,"Agriculture, Multidisciplinary",2196-5099,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,2196-5099
9413,BIOGEOCHEMICAL ADVANCES,NaN,3091-4493,SPRINGERNATURE,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",NaN,Geochemistry & Geophysics,3091-4493,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,3091-4493
9414,BULLETIN OF GEOPHYSICS AND OCEANOGRAPHY,2785-339X,2785-2970,IST NAZIONALE DI OCEANOGRAFIA E DI GEOFISICA,"BORGO GROTTA GIGANTE, 42-C, SGONICO, ITALY, TR...",English,Geochemistry & Geophysics,2785-339X,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,2785-339X


Let's see if we can get rid of anything with 100% title matching in the remaining missing 25 rows

In [63]:
m2 = m.loc[((m.id_issn_l.isna()) & (m._key_value.notna())), ['Journal title', 'ISSN', 'eISSN', 'Publisher name', 'Publisher address',
       'Languages', 'Web of Science Categories', 'ISSN_filled',]]
m2["journal_title"] = m2["Journal title"].str.lower().str.strip()
m2.merge(oa2, on="journal_title", how='left')

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_filled,journal_title,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat
0,ACTA PHYTOTAXONOMICA ET GEOBOTANICA,1346-7565,2189-7042,"JAPANESE SOC PLANT SYSTEMATICS, UNIV TOKYO, GR...","BOTANICAL GARDENS, 3-7-1 HAKUSAN, TOKYO, JAPAN...",English,Plant Sciences,1346-7565,acta phytotaxonomica et geobotanica,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,AGRARFORSCHUNG SCHWEIZ,NaN,2813-317X,AGRARFORSCHUNG,"AGROSCOPE LIEBEFELD-POSIEUX, POSIEUX, SWITZERL...",Multi-Language,"Agriculture, Multidisciplinary",2813-317X,agrarforschung schweiz,1663-7852,Agrarforschung Schweiz,None,None,None,https://openalex.org/S2490815337,"[1663-7852, 1663-7909]",2490815337,http://www.wikidata.org/entity/Q96325599,"1663-7852,1663-7909"
2,ANNALI DI BOTANICA,0365-0812,2239-3129,UNIV STUDI ROMA LA SAPIENZA,"PIAZZALE ALDO MORO, 5, ROME, ITALY, I-00185",Multi-Language,Plant Sciences | Environmental Sciences,0365-0812,annali di botanica,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,ANNALS ACADEMY OF MEDICINE SINGAPORE,NaN,2972-4066,ACAD MEDICINE SINGAPORE,"81 Kim Keat Road, #12-00 NKF Centre, Singapore...",English,"Medicine, General & Internal",2972-4066,annals academy of medicine singapore,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,ASIAN MYRMECOLOGY,1985-1944,NaN,UNIV MALAYSIA SABAH,"JALAN UMS, 88400, KOTA KINABALU, MALAYSIA, SAB...",English,Entomology,1985-1944,asian myrmecology,2462-2362,Asian Myrmecology,https://openalex.org/P4310322526,Universiti Malaysia Sabah,[https://openalex.org/P4310322526],https://openalex.org/S2482358896,[2462-2362],2482358896,None,2462-2362
5,ATW-INTERNATIONAL JOURNAL FOR NUCLEAR POWER,1431-5254,NaN,INFORUM VERLAGS-VERWALTUNGSGESELLSCHAFT MBH,"ATW-ADVERTISING-SUBSCRIPTIONS, ROBERT-KOCH-PLA...",Multi-Language,Nuclear Science & Technology,1431-5254,atw-international journal for nuclear power,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,AUSTRIAN JOURNAL OF FOREST SCIENCE,0379-5292,0375-524X,OSTERREICHISCHER AGRARVERLAG,"BANKGASSE 13, VIENNA 1, AUSTRIA, 1014",German,Forestry,0379-5292,austrian journal of forest science,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,BERICHTE UBER LANDWIRTSCHAFT,2196-5099,NaN,BUNDESMINISTERIUM ERNAHRUNG LANDWIRTSCHAFT,"DEICHMANNS AUE 29, BPMM, GERMANY, 53179",Multi-Language,"Agriculture, Multidisciplinary",2196-5099,berichte uber landwirtschaft,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,BIOGEOCHEMICAL ADVANCES,NaN,3091-4493,SPRINGERNATURE,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",NaN,Geochemistry & Geophysics,3091-4493,biogeochemical advances,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,BULLETIN OF GEOPHYSICS AND OCEANOGRAPHY,2785-339X,2785-2970,IST NAZIONALE DI OCEANOGRAFIA E DI GEOFISICA,"BORGO GROTTA GIGANTE, 42-C, SGONICO, ITALY, TR...",English,Geochemistry & Geophysics,2785-339X,bulletin of geophysics and oceanography,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [64]:
oa.loc[oa.id_mag=="2764689191", "host_organization_name"].iloc[0]

'Section Applied Botany of the German Botanical Society, German Society for Quality Research on Plant Foods (DGQ)'

we managed to match 2/19 with exact title match and one of them is wrong. so Unless we want to do it by hand, I wil leave these missing for now.

## SCOPUS

In [112]:
scopus = pd.read_excel("../data/citation_indexes/scopus/ext_list_May_2026.xlsx")
print(scopus.shape)
scopus.head()

(48580, 52)


,Sourcerecord ID,Source Title,ISSN,EISSN,Active or Inactive,Coverage,Titles Discontinued by Scopus,Article Language in Source (Three-Letter ISO Language Codes),Medline-sourced Title? (See additional details under separate tab.),Open Access Status,Articles in Press Included?,Added to List May 2026,Source Type,Title History Indication,Related Title 1,Other Related Title 2,Other Related Title 3,Other Related Title 4,Publisher,Publisher Imprints Grouped to Main Publisher,All Science Journal Classification Codes (ASJC),Top level:\n\nLife Sciences,Top level:\n\nSocial Sciences,Top level:\n\nPhysical Sciences,Top level:\n\nHealth Sciences,1000 \nGeneral,1100\nAgricultural and Biological Sciences,1200\nArts and Humanities,"1300\nBiochemistry, Genetics and Molecular Biology","1400\nBusiness, Management and Accounting",1500\nChemical Engineering,1600\nChemistry,1700\nComputer Science,1800\nDecision Sciences,1900\nEarth and Planetary Sciences,"2000\nEconomics, Econometrics and Finance",2100\nEnergy,2200\nEngineering,2300\nEnvironmental Science,2400\nImmunology and Microbiology,2500\nMaterials Science,2600\nMathematics,2700\nMedicine,2800\nNeuroscience,2900\nNursing,"3000\nPharmacology, Toxicology and Pharmaceutics",3100\nPhysics and Astronomy,3200\nPsychology,3300\nSocial Sciences,3400\nVeterinary,3500\nDentistry,3600\nHealth Professions
0,21101170611,@GRH,20349130,22959149,Active,2023-2024,NaN,ENG; FRE,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Association Francophone de Gestion des Relatio...,Association Francophone de Gestion des Relatio...,1407,NaN,Social Sciences,NaN,NaN,NaN,NaN,NaN,NaN,"Business, Management and Accounting",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,21100902951,[Hifu] Skin research,00181390,NaN,Inactive,1959-2001,NaN,JPN,NaN,NaN,NaN,NaN,Journal,Continued as,Skin Research,NaN,NaN,NaN,Osaka University Medical School,Osaka University Medical School,2708; 2725,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,26063,[Rinsho ketsueki] The Japanese journal of clin...,04851439,18820824,Inactive,1972-2016,NaN,NaN,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Nihon Rinsho Ketsueki Gakkai/Japan Society of ...,Nihon Rinsho Ketsueki Gakkai/Japan Society of ...,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,40160,[Zasshi] [Journal]. Nihon Kyobu Geka Gakkai,03694739,NaN,Inactive,1972-1997,NaN,JPN,NaN,NaN,NaN,NaN,Journal,Continued as,Japanese Journal of Thoracic and Cardiovascula...,NaN,NaN,NaN,Nihon Kyobu Geka Gakkai,Nihon Kyobu Geka Gakkai,2705,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,21100832755,1700-tal: Nordic Journal for Eighteenth-Centur...,16524772,20019866,Active,2019-2024; 2016-2017,NaN,DAN; FIN; GER; NOR,NaN,Unpaywall Open Access,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Swedish Society for Eighteenth-Century Studies,Swedish Society for Eighteenth-Century Studies,1213; 3316; 1202,NaN,Social Sciences,NaN,NaN,NaN,NaN,Arts and Humanities,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN


`len(ISSN)` should always be 8, then we can add - in the middle to match with openalex.

In [123]:
scopus["ISSN"].astype(str).apply(len).value_counts() 

ISSN
9    42306
4     6193
Name: count, dtype: int64

In [115]:
scopus["EISSN"].astype("str").apply(len).value_counts() 

EISSN
8    31621
3    16959
Name: count, dtype: int64

In [ ]:
scopus.ISSN.isna().sum() ## all three letters are because I made NaN a string

6271

In [76]:
scopus.ISSN.dropna().nunique() # !!

42306

In [ ]:
scopus["EISSN"].dropna().nunique() # !!

31618

In [ ]:
mask = scopus.ISSN.value_counts()[scopus.ISSN.value_counts() > 1]
scopus.loc[(scopus.ISSN.isin(mask.index))]

,Sourcerecord ID,Source Title,ISSN,EISSN,Active or Inactive,Coverage,Titles Discontinued by Scopus,Article Language in Source (Three-Letter ISO Language Codes),Medline-sourced Title? (See additional details under separate tab.),Open Access Status,Articles in Press Included?,Added to List May 2026,Source Type,Title History Indication,Related Title 1,Other Related Title 2,Other Related Title 3,Other Related Title 4,Publisher,Publisher Imprints Grouped to Main Publisher,All Science Journal Classification Codes (ASJC),Top level:\n\nLife Sciences,Top level:\n\nSocial Sciences,Top level:\n\nPhysical Sciences,Top level:\n\nHealth Sciences,1000 \nGeneral,1100\nAgricultural and Biological Sciences,1200\nArts and Humanities,"1300\nBiochemistry, Genetics and Molecular Biology","1400\nBusiness, Management and Accounting",1500\nChemical Engineering,1600\nChemistry,1700\nComputer Science,1800\nDecision Sciences,1900\nEarth and Planetary Sciences,"2000\nEconomics, Econometrics and Finance",2100\nEnergy,2200\nEngineering,2300\nEnvironmental Science,2400\nImmunology and Microbiology,2500\nMaterials Science,2600\nMathematics,2700\nMedicine,2800\nNeuroscience,2900\nNursing,"3000\nPharmacology, Toxicology and Pharmaceutics",3100\nPhysics and Astronomy,3200\nPsychology,3300\nSocial Sciences,3400\nVeterinary,3500\nDentistry,3600\nHealth Professions
18671,21101063000,"IEEE Advanced Information Management, Communic...",26932814,26932776,Inactive,2021,NaN,ENG,NaN,NaN,NaN,Added,Journal,NaN,NaN,NaN,NaN,NaN,Institute of Electrical and Electronics Engine...,IEEE,2105; 2102; 1802; 1710; 1705; 1405,NaN,Social Sciences,Physical Sciences,NaN,NaN,NaN,NaN,NaN,"Business, Management and Accounting",NaN,NaN,Computer Science,Decision Sciences,NaN,NaN,Energy,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18672,21101438394,"IEEE Advanced Information Management, Communic...",26932814,26932776,Active,2025,NaN,ENG,NaN,NaN,NaN,Added,Journal,NaN,NaN,NaN,NaN,NaN,Institute of Electrical and Electronics Engine...,IEEE,2105; 2102; 1802; 1710; 1705; 1405,NaN,Social Sciences,Physical Sciences,NaN,NaN,NaN,NaN,NaN,"Business, Management and Accounting",NaN,NaN,Computer Science,Decision Sciences,NaN,NaN,Energy,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
38007,21101041824,PRISM,25783491,25145347,Active,2019-2024,NaN,ENG,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Liverpool John Moores University,Liverpool John Moores University,3301; 3304,NaN,Social Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN
38008,21101396407,PRISM,25783491,25145347,Active,2022-2025,NaN,ENG,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Liverpool John Moores University,Liverpool John Moores University,3301; 3304,NaN,Social Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN
41025,21100227045,Revue Archéologique de l'Est,12667706,17607264,Active,2023-2024; 2021; 2019; 2011-2017,NaN,FRE,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Revue archeologique de l'Est,Revue archeologique de l'Est,3302; 1204,NaN,Social Sciences,NaN,NaN,NaN,NaN,Arts and Humanities,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN
41026,21101427907,Revue Archéologique de l'Est,12667706,17607264,Active,2024,NaN,FRE,NaN,NaN,NaN,Added,Journal,NaN,NaN,NaN,NaN,NaN,Revue archeologique de l'Est,Revue archeologique de l'Est,3302; 1204,NaN,Social Sciences,NaN,NaN,NaN,NaN,Arts and Humanities,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN


In [80]:
mask = scopus.EISSN.value_counts()[scopus.EISSN.value_counts() > 1]
scopus.loc[(scopus.EISSN.isin(mask.index))]

,Sourcerecord ID,Source Title,ISSN,EISSN,Active or Inactive,Coverage,Titles Discontinued by Scopus,Article Language in Source (Three-Letter ISO Language Codes),Medline-sourced Title? (See additional details under separate tab.),Open Access Status,Articles in Press Included?,Added to List May 2026,Source Type,Title History Indication,Related Title 1,Other Related Title 2,Other Related Title 3,Other Related Title 4,Publisher,Publisher Imprints Grouped to Main Publisher,All Science Journal Classification Codes (ASJC),Top level:\n\nLife Sciences,Top level:\n\nSocial Sciences,Top level:\n\nPhysical Sciences,Top level:\n\nHealth Sciences,1000 \nGeneral,1100\nAgricultural and Biological Sciences,1200\nArts and Humanities,"1300\nBiochemistry, Genetics and Molecular Biology","1400\nBusiness, Management and Accounting",1500\nChemical Engineering,1600\nChemistry,1700\nComputer Science,1800\nDecision Sciences,1900\nEarth and Planetary Sciences,"2000\nEconomics, Econometrics and Finance",2100\nEnergy,2200\nEngineering,2300\nEnvironmental Science,2400\nImmunology and Microbiology,2500\nMaterials Science,2600\nMathematics,2700\nMedicine,2800\nNeuroscience,2900\nNursing,"3000\nPharmacology, Toxicology and Pharmaceutics",3100\nPhysics and Astronomy,3200\nPsychology,3300\nSocial Sciences,3400\nVeterinary,3500\nDentistry,3600\nHealth Professions
18671,21101063000,"IEEE Advanced Information Management, Communic...",26932814,26932776,Inactive,2021,NaN,ENG,NaN,NaN,NaN,Added,Journal,NaN,NaN,NaN,NaN,NaN,Institute of Electrical and Electronics Engine...,IEEE,2105; 2102; 1802; 1710; 1705; 1405,NaN,Social Sciences,Physical Sciences,NaN,NaN,NaN,NaN,NaN,"Business, Management and Accounting",NaN,NaN,Computer Science,Decision Sciences,NaN,NaN,Energy,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18672,21101438394,"IEEE Advanced Information Management, Communic...",26932814,26932776,Active,2025,NaN,ENG,NaN,NaN,NaN,Added,Journal,NaN,NaN,NaN,NaN,NaN,Institute of Electrical and Electronics Engine...,IEEE,2105; 2102; 1802; 1710; 1705; 1405,NaN,Social Sciences,Physical Sciences,NaN,NaN,NaN,NaN,NaN,"Business, Management and Accounting",NaN,NaN,Computer Science,Decision Sciences,NaN,NaN,Energy,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
38007,21101041824,PRISM,25783491,25145347,Active,2019-2024,NaN,ENG,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Liverpool John Moores University,Liverpool John Moores University,3301; 3304,NaN,Social Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN
38008,21101396407,PRISM,25783491,25145347,Active,2022-2025,NaN,ENG,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Liverpool John Moores University,Liverpool John Moores University,3301; 3304,NaN,Social Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN
41025,21100227045,Revue Archéologique de l'Est,12667706,17607264,Active,2023-2024; 2021; 2019; 2011-2017,NaN,FRE,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Revue archeologique de l'Est,Revue archeologique de l'Est,3302; 1204,NaN,Social Sciences,NaN,NaN,NaN,NaN,Arts and Humanities,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN
41026,21101427907,Revue Archéologique de l'Est,12667706,17607264,Active,2024,NaN,FRE,NaN,NaN,NaN,Added,Journal,NaN,NaN,NaN,NaN,NaN,Revue archeologique de l'Est,Revue archeologique de l'Est,3302; 1204,NaN,Social Sciences,NaN,NaN,NaN,NaN,Arts and Humanities,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN


seems like we can keep the last, we don't care about active dates and everything else looks the same. Are there any rows where both ISSN and EISSN are emtpy?

In [85]:
print(scopus[(scopus.ISSN.isna()) & (scopus.EISSN.isna())].shape)
scopus[(scopus.ISSN.isna()) & (scopus.EISSN.isna())].sample(10)

(79, 52)


,Sourcerecord ID,Source Title,ISSN,EISSN,Active or Inactive,Coverage,Titles Discontinued by Scopus,Article Language in Source (Three-Letter ISO Language Codes),Medline-sourced Title? (See additional details under separate tab.),Open Access Status,Articles in Press Included?,Added to List May 2026,Source Type,Title History Indication,Related Title 1,Other Related Title 2,Other Related Title 3,Other Related Title 4,Publisher,Publisher Imprints Grouped to Main Publisher,All Science Journal Classification Codes (ASJC),Top level:\n\nLife Sciences,Top level:\n\nSocial Sciences,Top level:\n\nPhysical Sciences,Top level:\n\nHealth Sciences,1000 \nGeneral,1100\nAgricultural and Biological Sciences,1200\nArts and Humanities,"1300\nBiochemistry, Genetics and Molecular Biology","1400\nBusiness, Management and Accounting",1500\nChemical Engineering,1600\nChemistry,1700\nComputer Science,1800\nDecision Sciences,1900\nEarth and Planetary Sciences,"2000\nEconomics, Econometrics and Finance",2100\nEnergy,2200\nEngineering,2300\nEnvironmental Science,2400\nImmunology and Microbiology,2500\nMaterials Science,2600\nMathematics,2700\nMedicine,2800\nNeuroscience,2900\nNursing,"3000\nPharmacology, Toxicology and Pharmaceutics",3100\nPhysics and Astronomy,3200\nPsychology,3300\nSocial Sciences,3400\nVeterinary,3500\nDentistry,3600\nHealth Professions
24390,4800154020,Journal of cardiology. Supplement,NaN,NaN,Inactive,1987-1992,NaN,NaN,NaN,NaN,NaN,NaN,Journal,Incorporated into,Journal of Cardiology,NaN,NaN,NaN,Japanese College of Cardiology (Nippon-Sinzoby...,Japanese College of Cardiology (Nippon-Sinzoby...,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
28958,19500157727,Journal of the American Physical Therapy Assoc...,NaN,NaN,Inactive,1962-1963,NaN,NaN,NaN,NaN,NaN,NaN,Journal,Continued as,Physical Therapy,NaN,NaN,NaN,American Physical Therapy Association,American Physical Therapy Association,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
11539,100588,Dental news,NaN,NaN,Inactive,1968-1969,NaN,NaN,NaN,NaN,NaN,NaN,Journal,Incorporated into,DP. Dental practice,NaN,NaN,NaN,Dental Publications,Dental Publications,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19142,93988,Images; Marquette University dental reflections,NaN,NaN,Inactive,1965-1967,NaN,NaN,NaN,NaN,NaN,NaN,Journal,Continued as,Dental images,NaN,NaN,NaN,Marquette University Press,Marquette University Press,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
11652,88174,Deutsche Schwesternzeitung,NaN,NaN,Inactive,1965-1970,NaN,NaN,NaN,NaN,NaN,NaN,Journal,Continued as,Deutsche Krankenpflegezeitschrift,NaN,NaN,NaN,Kohlhammer,Kohlhammer,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
36549,18500161300,Performance Textiles,NaN,NaN,Inactive,2004-2008,NaN,NaN,NaN,NaN,NaN,NaN,Trade Journal,Formerly known as,Industrial Textiles,NaN,NaN,NaN,NaN,NaN,2209; 2507; 1403,NaN,Social Sciences,Physical Sciences,NaN,NaN,NaN,NaN,NaN,"Business, Management and Accounting",NaN,NaN,NaN,NaN,NaN,NaN,NaN,Engineering,NaN,NaN,Materials Science,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
45205,71348,The Admitting management journal,NaN,NaN,Inactive,1988-1990,NaN,NaN,NaN,NaN,NaN,NaN,Journal,Continued as,The NAHAM management journal,NaN,NaN,NaN,National Association of Hospital Admitting Man...,National Association of Hospital Admitting Man...,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3020,50063,Annali Sclavo. Collana monografica,NaN,NaN,Inacti

In [86]:
scopus[(scopus.ISSN.isna()) & (scopus.EISSN.isna())]["Active or Inactive"].value_counts()

Active or Inactive
Inactive    78
Active       1
Name: count, dtype: int64

they are mostly very old and inactive.. I'm ok with dropping those as well..

In [117]:
print(scopus.shape)
scopus = scopus.drop_duplicates(subset=['ISSN', "EISSN"], keep="last")
print(scopus.shape)

(48580, 52)
(48499, 52)


In [118]:
scopus.ISSN = scopus.ISSN.astype("string")
scopus.EISSN = scopus.EISSN.astype("string")

In [103]:
scopus.head()

,Sourcerecord ID,Source Title,ISSN,EISSN,Active or Inactive,Coverage,Titles Discontinued by Scopus,Article Language in Source (Three-Letter ISO Language Codes),Medline-sourced Title? (See additional details under separate tab.),Open Access Status,Articles in Press Included?,Added to List May 2026,Source Type,Title History Indication,Related Title 1,Other Related Title 2,Other Related Title 3,Other Related Title 4,Publisher,Publisher Imprints Grouped to Main Publisher,All Science Journal Classification Codes (ASJC),Top level:\n\nLife Sciences,Top level:\n\nSocial Sciences,Top level:\n\nPhysical Sciences,Top level:\n\nHealth Sciences,1000 \nGeneral,1100\nAgricultural and Biological Sciences,1200\nArts and Humanities,"1300\nBiochemistry, Genetics and Molecular Biology","1400\nBusiness, Management and Accounting",1500\nChemical Engineering,1600\nChemistry,1700\nComputer Science,1800\nDecision Sciences,1900\nEarth and Planetary Sciences,"2000\nEconomics, Econometrics and Finance",2100\nEnergy,2200\nEngineering,2300\nEnvironmental Science,2400\nImmunology and Microbiology,2500\nMaterials Science,2600\nMathematics,2700\nMedicine,2800\nNeuroscience,2900\nNursing,"3000\nPharmacology, Toxicology and Pharmaceutics",3100\nPhysics and Astronomy,3200\nPsychology,3300\nSocial Sciences,3400\nVeterinary,3500\nDentistry,3600\nHealth Professions
0,21101170611,@GRH,20349130,22959149,Active,2023-2024,NaN,ENG; FRE,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Association Francophone de Gestion des Relatio...,Association Francophone de Gestion des Relatio...,1407,NaN,Social Sciences,NaN,NaN,NaN,NaN,NaN,NaN,"Business, Management and Accounting",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,21100902951,[Hifu] Skin research,00181390,NaN,Inactive,1959-2001,NaN,JPN,NaN,NaN,NaN,NaN,Journal,Continued as,Skin Research,NaN,NaN,NaN,Osaka University Medical School,Osaka University Medical School,2708; 2725,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,26063,[Rinsho ketsueki] The Japanese journal of clin...,04851439,18820824,Inactive,1972-2016,NaN,NaN,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Nihon Rinsho Ketsueki Gakkai/Japan Society of ...,Nihon Rinsho Ketsueki Gakkai/Japan Society of ...,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,40160,[Zasshi] [Journal]. Nihon Kyobu Geka Gakkai,03694739,NaN,Inactive,1972-1997,NaN,JPN,NaN,NaN,NaN,NaN,Journal,Continued as,Japanese Journal of Thoracic and Cardiovascula...,NaN,NaN,NaN,Nihon Kyobu Geka Gakkai,Nihon Kyobu Geka Gakkai,2705,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,21100832755,1700-tal: Nordic Journal for Eighteenth-Centur...,16524772,20019866,Active,2019-2024; 2016-2017,NaN,DAN; FIN; GER; NOR,NaN,Unpaywall Open Access,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Swedish Society for Eighteenth-Century Studies,Swedish Society for Eighteenth-Century Studies,1213; 3316; 1202,NaN,Social Sciences,NaN,NaN,NaN,NaN,Arts and Humanities,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN


In [119]:
scopus[["ISSN", "EISSN"]].isna().sum()

ISSN      6193
EISSN    16881
dtype: int64

In [120]:
scopus["ISSN"] = scopus["ISSN"].str.slice(0,4) + "-" + scopus["ISSN"].str.slice(4,8)
scopus["EISSN"] = scopus["EISSN"].str.slice(0,4) + "-" + scopus["EISSN"].str.slice(4,8)

scopus[["ISSN", "EISSN"]].isna().sum()

ISSN      6193
EISSN    16881
dtype: int64

In [122]:
scopus[["ISSN", "EISSN"]].sample(5)

,ISSN,EISSN
18473,0272-457X,<NA>
26206,0970-9274,2456-6608
4243,2223-4594,<NA>
20757,1556-8881,1556-8873
45024,0212-4408,1988-7868


In [124]:
scopus["ISSN_all"] = scopus[["ISSN", "EISSN"]].apply(lambda r: [v for v in r.tolist() if pd.notna(v)], axis=1)
scopus["ISSN_filled"] = scopus.ISSN
scopus.loc[scopus.ISSN_filled.isna(), "ISSN_filled"] = scopus.loc[scopus.ISSN_filled.isna(), "EISSN"]

scopus.head()

,Sourcerecord ID,Source Title,ISSN,EISSN,Active or Inactive,Coverage,Titles Discontinued by Scopus,Article Language in Source (Three-Letter ISO Language Codes),Medline-sourced Title? (See additional details under separate tab.),Open Access Status,Articles in Press Included?,Added to List May 2026,Source Type,Title History Indication,Related Title 1,Other Related Title 2,Other Related Title 3,Other Related Title 4,Publisher,Publisher Imprints Grouped to Main Publisher,All Science Journal Classification Codes (ASJC),Top level:\n\nLife Sciences,Top level:\n\nSocial Sciences,Top level:\n\nPhysical Sciences,Top level:\n\nHealth Sciences,1000 \nGeneral,1100\nAgricultural and Biological Sciences,1200\nArts and Humanities,"1300\nBiochemistry, Genetics and Molecular Biology","1400\nBusiness, Management and Accounting",1500\nChemical Engineering,1600\nChemistry,1700\nComputer Science,1800\nDecision Sciences,1900\nEarth and Planetary Sciences,"2000\nEconomics, Econometrics and Finance",2100\nEnergy,2200\nEngineering,2300\nEnvironmental Science,2400\nImmunology and Microbiology,2500\nMaterials Science,2600\nMathematics,2700\nMedicine,2800\nNeuroscience,2900\nNursing,"3000\nPharmacology, Toxicology and Pharmaceutics",3100\nPhysics and Astronomy,3200\nPsychology,3300\nSocial Sciences,3400\nVeterinary,3500\nDentistry,3600\nHealth Professions,ISSN_all,ISSN_filled
0,21101170611,@GRH,2034-9130,2295-9149,Active,2023-2024,NaN,ENG; FRE,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Association Francophone de Gestion des Relatio...,Association Francophone de Gestion des Relatio...,1407,NaN,Social Sciences,NaN,NaN,NaN,NaN,NaN,NaN,"Business, Management and Accounting",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"[2034-9130, 2295-9149]",2034-9130
1,21100902951,[Hifu] Skin research,0018-1390,<NA>,Inactive,1959-2001,NaN,JPN,NaN,NaN,NaN,NaN,Journal,Continued as,Skin Research,NaN,NaN,NaN,Osaka University Medical School,Osaka University Medical School,2708; 2725,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,[0018-1390],0018-1390
2,26063,[Rinsho ketsueki] The Japanese journal of clin...,0485-1439,1882-0824,Inactive,1972-2016,NaN,NaN,NaN,NaN,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Nihon Rinsho Ketsueki Gakkai/Japan Society of ...,Nihon Rinsho Ketsueki Gakkai/Japan Society of ...,2700,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"[0485-1439, 1882-0824]",0485-1439
3,40160,[Zasshi] [Journal]. Nihon Kyobu Geka Gakkai,0369-4739,<NA>,Inactive,1972-1997,NaN,JPN,NaN,NaN,NaN,NaN,Journal,Continued as,Japanese Journal of Thoracic and Cardiovascula...,NaN,NaN,NaN,Nihon Kyobu Geka Gakkai,Nihon Kyobu Geka Gakkai,2705,NaN,NaN,NaN,Health Sciences,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Medicine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,[0369-4739],0369-4739
4,21100832755,1700-tal: Nordic Journal for Eighteenth-Centur...,1652-4772,2001-9866,Active,2019-2024; 2016-2017,NaN,DAN; FIN; GER; NOR,NaN,Unpaywall Open Access,NaN,NaN,Journal,NaN,NaN,NaN,NaN,NaN,Swedish Society for Eighteenth-Century Studies,Swedish Society for Eighteenth-Century Studies,1213; 3316; 1202,NaN,Social Sciences,NaN,NaN,NaN,NaN,Arts and Humanities,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Social Sciences,NaN,NaN,NaN,"[1652-4772, 2001-9866]",1652-4772


In [130]:
print("Left shape:  ", scopus.shape)
print("Right shape: ", oa.shape)
m, s = merge_by_bidirectional_preference(scopus, oa, left_id_col='ISSN_filled', left_list_col='ISSN_all',
                                      right_id_col='id_issn_l', right_list_col='id_issn')
print("Merged shape:", m.shape)
cols_of_interest = ['Source Title', 'ISSN', 'EISSN','ISSN_filled', 'id_issn_l', 'display_name',
       'host_organization', 'host_organization_name',
       'host_organization_lineage', 'id_openalex', 'id_mag', 'id_wikidata',
       'id_issn_concat', 'journal_title', 'l_pref', 'r_pref', 'score',
       '_key_value']
m[cols_of_interest].isna().sum() / m.shape[0]

Left shape:   (48499, 54)
Right shape:  (166270, 11)
Merged shape: (48499, 67)


Source Title                 0.000000
ISSN                         0.127693
EISSN                        0.348069
ISSN_filled                  0.000021
id_issn_l                    0.128415
display_name                 0.128415
host_organization            0.326089
host_organization_name       0.326192
host_organization_lineage    0.326027
id_openalex                  0.128415
id_mag                       0.128415
id_wikidata                  0.420318
id_issn_concat               0.128415
journal_title                0.128415
l_pref                       0.000000
r_pref                       0.128415
score                        0.128415
_key_value                   0.000021
dtype: float64

In [131]:
s

{'total_left_rows': 48499,
 'matched_rows': 42271,
 'unmatched_rows': 6228,
 'matched_by_l_pref': l_pref
 0.0    42117
 1.0      154
 Name: count, dtype: int64,
 'matched_by_r_pref': r_pref
 0.0    40495
 1.0     1692
 2.0       76
 3.0        7
 4.0        1
 Name: count, dtype: int64}

In [132]:
6228/48499

0.12841501886636839